### PR TITLE
feat: userIdMapping

### PR DESCRIPTION
**Pull request by jscyo**
## Summary of change

Adds userIdMapping recipe in the backend SDK

## Related issues

-  https://github.com/supertokens/supertokens-node/issues/198

## Checklist for important updates

-   [x] Changelog has been updated
-   [x] `coreDriverInterfaceSupported.json` file has been updated (if needed)
    -   Along with the associated array in `lib/ts/version.ts`
-   [x] `frontendDriverInterfaceSupported.json` file has been updated (if needed)
-   [ ] Changes to the version if needed
    -   In `package.json`
    -   In `package-lock.json`
    -   In `lib/ts/version.ts`
-   [x] Had run `npm run build-pretty`
-   [x] Had installed and ran the pre-commit hook
-   [ ] Issue this PR against the latest non released version branch.
    -   To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    -   If no such branch exists, then create one from the latest released branch.
-   [ ] If have added a new web framework, update the `add-ts-no-check.js` file to include that
-   [ ] If added a new recipe / api interface, then make sure that the implementation of it uses NON arrow functions only (like `someFunc: function () {..}`).
-   [ ] If added a new recipe, then make sure to expose it inside the recipe folder present in the root of this repo. We also need to expose its types.

## Remaining TODOs for this PR

-   [x] deleteUser, userPagination, third-party, passwordless need to be changed

- [x] deleteUser: if userIdMapping exists: do two delete user calls to the core, one to delete with superTokensUserId and one to delete with  externalUserId
- [x] userPagination: will need a rate-limited API which can get the userIdMappings
- [ ] Remove userIdMapping recipe, shift functions to the superTokens level similar to userPagination and delete User

### CODE REVIEW

**Comment by jscyo** [comment]
## Feedback
EmailPassword recipe feedback:
- [ ] SignIn function should also call recipe mapping
- [ ] userContext should be passed to the recipe mapping functions
- [ ] we should store userIdMapping in userContext, so if multiple apis are calling each other there wont be duplicate calls
- [ ] the userIdMapping recipe functions should check the userContext to reduce clutter when the functions are being called in the recipe
- [ ] resetPasswordUsingToken function does should also call userIdMapping
- [ ] Implement userId mapping in emailverification functions

File: lib/ts/recipe/useridmapping/index.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            userContext: userContext === undefined ? {} : userContext,
+        });
+    }
+}
+
+export enum UserIdType {
+    SUPERTOKENS,
+    EXTERNAL,
+    ANY,
+}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/index.ts:50]: No need for this enum type

File: lib/ts/recipe/useridmapping/utils.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    if (userIdType === UserIdType.SUPERTOKENS) {
+        return "SUPERTOKENS";
+    }
+
+    if (userIdType === UserIdType.EXTERNAL) {
+        return "EXTERNAL";
+    }
+
+    return "ANY";
+}
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/utils.ts:47]: No need

File: lib/ts/recipe/emailpassword/recipeImplementation.ts [hunk 3]
```diff
@@ -112,14 +179,35 @@ export default function getRecipeInterface(querier: Querier): RecipeInterface {
                 token,
                 newPassword,
             });
+
+            if (response.status === "OK") {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailpassword/recipeImplementation.ts:183]:
```suggestion
            if (response.status === "OK" && response.userId !== undefined) {
```

File: lib/ts/recipe/useridmapping/types.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+              externalUserIdInfo: string | undefined;
+          }
+        | {
+              status: "UNKNOWN_MAPPING_ERROR";
+          }
+    >;
+
+    deleteUserIdMapping: (input: {
+        userId: string;
+        userIdType?: "SUPERTOKENS" | "EXTERNAL" | "ANY";
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/types.ts:75]: userContext missing

File: lib/ts/recipe/useridmapping/types.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    }) => Promise<{
+        status: "OK";
+        didMappingExist: boolean;
+    }>;
+
+    updateOrDeleteUserIdMappingInfo: (input: {
+        userId: string;
+        userIdType?: "SUPERTOKENS" | "EXTERNAL" | "ANY";
+        externalUserIdInfo: string | null;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/types.ts:84]: userContext missing

File: lib/ts/recipe/useridmapping/index.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        });
+    }
+
+    static async getUserIdMapping(userId: string, userIdType: UserIdType, userContext?: any) {
+        return await Recipe.getInstanceOrThrowError().recipeInterfaceImpl.getUserIdMapping({
+            userId,
+            userIdType: getUserIdTypeAsString(userIdType),
+            userContext: userContext === undefined ? {} : userContext,
+        });
+    }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/index.ts:43]: need to expose all the recipe interface functions

File: lib/ts/recipe/emailverification/recipeImplementation.ts [hunk 7]
```diff
@@ -1,22 +1,34 @@
...
         createEmailVerificationToken: async function ({
             userId,
             email,
+            userContext,
         }: {
             userId: string;
             email: string;
+            userContext: any;
         }): Promise<
             | {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/emailverification/recipeImplementation.ts:17]: Undo all the changes for email verificaiton recipe

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

File: lib/ts/recipe/useridmapping/recipe.ts [hunk 8]
```diff
@@ -26,6 +26,8 @@ import { RecipeInterface, TypeInput, TypeNormalisedInput } from "./types";
 import { validateAndNormaliseUserInput } from "./utils";
 import OverrideableBuilder from "supertokens-js-override";
 
+let isUserIdMappingRecipeInitialized = false;
```
Comment (rishabhpoddar) [inline lib/ts/recipe/useridmapping/recipe.ts:29]: You don't need this.. just do Recipe.instance === undefined instead

File: lib/ts/recipe/passwordless/recipeImplementation.ts [hunk 9]
```diff
@@ -17,6 +19,20 @@ export default function getRecipeInterface(querier: Querier): RecipeInterface {
                 new NormalisedURLPath("/recipe/signinup/code/consume"),
                 copyAndRemoveUserContext(input)
             );
+
+            if (isUserIdMappingRecipeInitialized) {
+                if (isUserIdMappingRecipeInitialized) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/passwordless/recipeImplementation.ts]: why twice?

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 10]
```diff
@@ -1,34 +1,60 @@
...
             };
+            userContext: any;
         }): Promise<{ status: "OK"; createdNewUser: boolean; user: User }> {
             let response = await querier.sendPostRequest(new NormalisedURLPath("/recipe/signinup"), {
                 thirdPartyId,
                 thirdPartyUserId,
                 email,
             });
+
+            if (!response.createdNewUser) {
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/recipeImplementation.ts:29]: Remove this check

File: lib/ts/recipe/thirdparty/recipeImplementation.ts [hunk 11]
```diff
@@ -1,34 +1,60 @@
...
+        }: {
+            userId: string;
+            userContext: any;
+        }): Promise<User | undefined> {
+            if (isUserIdMappingRecipeInitialized) {
+                let userIdMappingResponse = await getUserIdMapping(userId, "ANY", userContext);
+                if (userIdMappingResponse.status === "OK") {
+                    userId = userIdMappingResponse.superTokensUserId;
+                }
+            }
```
Comment (rishabhpoddar) [inline lib/ts/recipe/thirdparty/recipeImplementation.ts:57]: You have not mapped the external userID back

File: lib/ts/supertokens.ts [hunk 12]
```diff
@@ -200,6 +202,89 @@ export default class SuperTokens {
...
+                            try {
+                                const userIdMappingResponse = await getUserIdMapping(
+                                    userObj.user.id,
+                                    "SUPERTOKENS",
+                                    undefined
+                                );
+
+                                if (userIdMappingResponse.status === "OK") {
+                                    updatedUsersArray[i] = {
+                                        recipeId: userObj.recipeId,
```
Comment (rishabhpoddar) [inline lib/ts/supertokens.ts:229]:
```suggestion
                                        ...userObj,
```

File: lib/ts/supertokens.ts [hunk 13]
```diff
@@ -200,6 +202,89 @@ export default class SuperTokens {
...
+                                // Something went wrong when fetching userId mapping data
+                                reject(e);
+                            }
+                        })
+                );
+
+                let userIdMapping = await getUserIdMapping(response.users[i].id, "SUPERTOKENS", undefined);
+                if (userIdMapping.status === "OK") {
+                    response.users[i].id = userIdMapping.externalUserId;
+                }
```
Comment (rishabhpoddar) [inline lib/ts/supertokens.ts:249]: What is this for?

File: lib/ts/supertokens.ts [hunk 14]
```diff
@@ -211,9 +296,22 @@ export default class SuperTokens {
...
+                let userIdMappingResponse = await getUserIdMapping(input.userId, "ANY", undefined);
+                if (userIdMappingResponse.status === "OK") {
+                    await querier.sendPostRequest(new NormalisedURLPath("/user/remove"), {
+                        userId: userIdMappingResponse.superTokensUserId,
+                    });
+
+                    await querier.sendPostRequest(new NormalisedURLPath("/user/remove"), {
+                        userId: userIdMappingResponse.externalUserId,
+                    });
+                }
```
Comment (rishabhpoddar) [inline lib/ts/supertokens.ts:309]: else???

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: lib/ts/supertokens.ts [hunk 15]
```diff
@@ -296,9 +291,13 @@ export default class SuperTokens {
         let cdiVersion = await querier.getAPIVersion();
         if (maxVersion("2.10", cdiVersion) === cdiVersion) {
             // delete user is only available >= CDI 2.10
-            if (isUserIdMappingRecipeInitialized) {
+
+            let doesUserIdMappingExist = false;
```
Comment (rishabhpoddar) [inline lib/ts/supertokens.ts:295]: Call this variable somethign better

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**